### ISSUE TITLE
Pylint not showing errors in sublime 3

### ISSUE THREAD

**Issue by auraz** [biermeester/Pylinter#44]
Similar to recent issue: https://github.com/biermeester/Pylinter/issues/13
Example:

```
Writing file /Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py with encoding UTF-8 (atomic)
 - PyLinter: Verbose is True
 - PyLinter: Running Pylinter on /Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py
 - PyLinter: Current PYTHONPATH is '/Users/kry/mitx_all/edx-platform/common/lib:/Users/kry/mitx_all/edx-platform/lms/djangoapps:/Users/kry/mitx_all/edx-platform/cms/djangoapps:/Users/kry/mitx_all/edx-platform/common/djangoapps:/Users/kry/mitx_all/edx-platform/lms/lib:/Users/kry/mitx_all/edx-platform/cms/lib'
 - PyLinter: Updated PYTHONPATH is '/Users/kry/mitx_all/edx-platform/common/lib:/Users/kry/mitx_all/edx-platform/lms/djangoapps:/Users/kry/mitx_all/edx-platform/cms/djangoapps:/Users/kry/mitx_all/edx-platform/common/djangoapps:/Users/kry/mitx_all/edx-platform/lms/lib:/Users/kry/mitx_all/edx-platform/cms/lib'
 - PyLinter: Running command with Pylint (0, 28, 0)
 - PyLinter: /Users/kry/mitx_all/python/bin/python /Users/kry/mitx_all/python/lib/python2.7/site-packages/pylint/lint.py --output-format=parseable --include-ids=y --rcfile=/Users/kry/mitx_all/edx-platform/pylintrc /Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py
reloading plugin Pylinter.pylinter
 - PyLinter: Pylint executable *not* found
 - PyLinter: Seaching for lint.py module...
 - PyLinter: Pylint path /Library/Python/2.7/site-packages/pylint/lint.py found
 - PyLinter: Pylint version (0, 28, 0) found
 - PyLinter: No errors found

```

While when running command separately:

```
$ /Users/kry/mitx_all/python/bin/python /Users/kry/mitx_all/python/lib/python2.7/site-packages/pylint/lint.py --output-format=parseable --include-ids=y --rcfile=/Users/kry/mitx_all/edx-platform/pylintrc /Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py
/Users/kry/Library/Application:1: [F0001(fatal)] No module named /Users/kry/Library/Application
Support/Sublime:1: [F0001(fatal)] No module named Support/Sublime
Text:1: [F0001(fatal)] No module named Text
3/Packages/Pylinter/pylinter.py:1: [F0001(fatal)] No module named 3/Packages/Pylinter/pylinter.py
```

Investigation shows that regexp P_PYLINT_ERROR for path `PYLINT_VERSION[0] == 0:` (I use pylint 0.28) was incorrect:

```
In [179]: %paste
P_PYLINT_ERROR = re.compile(r"""
            ^(?P<file>.+?):(?P<line>[0-9]+):\ # file name and line number
            \[(?P<type>[a-z])(?P<errno>\d+)   # message type and error number
                                              # e.g. E0101
            (,\ (?P<hint>.+))?\]\             # optional class or function name
            (?P<msg>.*)                       # finally, the error message
            """, re.IGNORECASE | re.VERBOSE)

## -- End pasted text --

In [180]: P_PYLINT_ERROR
Out[180]:
re.compile(r'\n            ^(?P<file>.+?):(?P<line>[0-9]+):\ # file name and line number\n            \[(?P<type>[a-z])(?P<errno>\d+)   # message type and error number\n                                              # e.g. E0101\n            (,\ (?P<hint>.+))?\]\             # optional class or function name\n            (?P<msg>.*)                       # finally, the error message\n            ',
re.IGNORECASE|re.VERBOSE)

In [181]: x = """/Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py:18: [F0401(import-error)] Unable to import 'sublime'"""

In [182]: print x
/Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py:18: [F0401(import-error)] Unable to import 'sublime'

In [183]: re.search(P_PYLINT_ERROR, x)
```

After changing it to:

```
P_PYLINT_ERROR = re.compile(r"""
            ^(?P<file>.+?):(?P<line>[0-9]+):\s*\ # file name and line number
            \[(?P<type>[a-z\(\)-])(?P<errno>\d+)   # message type and error number
                                              # e.g. E0101
            (?P<hint>.+)?\]\             # optional class or function name
            (?P<msg>.*)                       # finally, the error message
            """, re.IGNORECASE | re.VERBOSE | re.DOTALL)
```

All starts to work:

```
Writing file /Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py with encoding UTF-8 (atomic)
 - PyLinter: Verbose is True
 - PyLinter: Running Pylinter on /Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py
 - PyLinter: Current PYTHONPATH is '/Users/kry/mitx_all/edx-platform/common/lib:/Users/kry/mitx_all/edx-platform/lms/djangoapps:/Users/kry/mitx_all/edx-platform/cms/djangoapps:/Users/kry/mitx_all/edx-platform/common/djangoapps:/Users/kry/mitx_all/edx-platform/lms/lib:/Users/kry/mitx_all/edx-platform/cms/lib'
 - PyLinter: Updated PYTHONPATH is '/Users/kry/mitx_all/edx-platform/common/lib:/Users/kry/mitx_all/edx-platform/lms/djangoapps:/Users/kry/mitx_all/edx-platform/cms/djangoapps:/Users/kry/mitx_all/edx-platform/common/djangoapps:/Users/kry/mitx_all/edx-platform/lms/lib:/Users/kry/mitx_all/edx-platform/cms/lib'
 - PyLinter: Running command with Pylint (0, 28, 0)
 - PyLinter: /Users/kry/mitx_all/python/bin/python /Users/kry/mitx_all/python/lib/python2.7/site-packages/pylint/lint.py --output-format=parseable --include-ids=y --rcfile=/Users/kry/mitx_all/edx-platform/pylintrc /Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py
reloading plugin Pylinter.pylinter
 - PyLinter: Pylint executable *not* found
 - PyLinter: Seaching for lint.py module...
 - PyLinter: Pylint path /Library/Python/2.7/site-packages/pylint/lint.py found
 - PyLinter: Pylint version (0, 28, 0) found
 - PyLinter: F0401: Unable to import 'sublime' 
 - PyLinter: F0401: Unable to import 'sublime_plugin' 
 - PyLinter: W0603: Using the global statement 
 - PyLinter: C0111: Missing docstring 
 - PyLinter: C0111: Missing docstring 
 - PyLinter: C0111: Missing docstring 
 - PyLinter: C0111: Missing docstring 
 - PyLinter: W0603: Using the global statement 
 - PyLinter: C0103: Invalid name "p" for type variable (should match [a-z_][a-z0-9_]{2,30}$) 
 - PyLinter: C0111: Missing docstring 
 - PyLinter: C0111: Missing docstring 
 - PyLinter: W0602: Using global for 'PYLINTER_ERRORS' but no assignment is done 
 - PyLinter: C0103: Invalid name "p" for type variable (should match [a-z_][a-z0-9_]{2,30}$) 
 - PyLinter: C0103: Invalid name "m" for type variable (should match [a-z_][a-z0-9_]{2,30}$) 
 - PyLinter: C0111: Missing docstring 
 - PyLinter: W0603: Using the global statement 
```

**Comment by biermeester** [comment]
Maybe you just write very clean code? ;)

Could you give me some more info please? OS, log output, settings etc.

**Comment by auraz** [comment]
Sure, give me 5 mins

**Comment by biermeester** [comment]
Ah, thanks. I'll have a look.

**Comment by auraz** [comment]
Done

**Comment by biermeester** [comment]
I hope I got it right. Your problem was that, for some weird reason, your pylint output contained newlines that weren't matched in the original regular expression and adding `re.DOTALL` solved that?

Anyway, I updated the regular expressions. Let me know how the new version works for you.

Thanks for the feedback!

**Comment by auraz** [comment]
I think problem was in / in file path and space after : before [. Will test now. I've tested old regexp on single line, so this is not new line problem.

**Comment by auraz** [comment]
I've installed Pylinter via package control, but I can't upgrade pylinter via package control - pylinter not in list of upgradable packages.

**Comment by biermeester** [comment]
I think that there's a delay before Package Control picks up on an update.

**Comment by auraz** [comment]
Found another issue:
for errors like 

```
 [W1401(anomalous-backslash-in-string)] Anomalous backslash in string: \'\\w\'. String constant might be missing an r prefix.\n
```

pylinter produces error

```
 - PyLinter: /Users/kry/mitx_all/python/bin/python /Users/kry/mitx_all/python/lib/python2.7/site-packages/pylint/lint.py --output-format=parseable --include-ids=y --rcfile=/Users/kry/mitx_all/edx-platform/pylintrc /Users/kry/mitx_all/edx-platform/common/lib/capa/capa/tests/test_responsetypes.py
Exception in thread Thread-84:
Traceback (most recent call last):
  File "X/threading.py", line 639, in _bootstrap_inner
  File "/Users/kry/Library/Application Support/Sublime Text 3/Packages/Pylinter/pylinter.py", line 494, in run
    lines = [line for line in output.decode().split('\n')]  # pylint: disable=E1103
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc3 in position 844: invalid continuation byte
```

**Comment by auraz** [comment]
Checked - new version of regexp do not work.

**Comment by biermeester** [comment]
Darn. I reverted the last commit. I'll try and have a look at it again later, using similar paths as you provided in your first post.

**Comment by iambibhas** [comment]
I just installed and received this - 

```
Traceback (most recent call last):
  File "/opt/sublime_text/sublime_plugin.py", line 549, in run_
    return self.run(edit)
  File "/home/bibhas/.config/sublime-text-3/Packages/Pylinter/pylinter.py", line 267, in run
    settings = PylSet.read_settings()
  File "/home/bibhas/.config/sublime-text-3/Packages/Pylinter/pylinter.py", line 151, in read_settings
    if PYLINT_VERSION[0] != 0:
TypeError: 'NoneType' object is not subscriptable
```

**Comment by biermeester** [comment]
Can you give some more info? Here it runs fine under Linux/ST3.

**Comment by iambibhas** [comment]
Never mind. Didn't have PyLint installed. ST should open the README file of all plugins after installing and all the relevant instructions should be there.

**Comment by biermeester** [comment]
Ok, I'll see if my plugin can give a nice warning if Pylint is not installed. Thanks for the feedback.

**Comment by jlujan** [comment]
I am having similar issues in Sublime Text 2. I am on a fresh install of OS X Mavericks and my python environment is provided by Anaconda.

``` pre
a-mbp:~ user$ which python
/Users/user/anaconda/bin/python
a-mbp:~ user$ python --version
Python 2.7.6 :: Anaconda 1.8.0 (x86_64)
a-mbp:~ user$ which pylint
/Users/atrapp/anaconda/bin/pylint
a-mbp:~ user$ pylint --version
No config file found, using default configuration
pylint 1.1.0, 
astroid 1.0.1, common 0.60.1
Python 2.7.6 |Anaconda 1.8.0 (x86_64)| (default, Jan 10 2014, 11:23:15) 
[GCC 4.0.1 (Apple Inc. build 5493)]
```

I now get this with the plugin

``` code
Reloading plugin Users/user//Library/Application Support/Sublime Text 2/Packages/Pylinter/pylinter.py
 - PyLinter: Pylint executable *not* found
 - PyLinter: Seaching for lint.py module...
Traceback (most recent call last):
  File "./sublime_plugin.py", line 62, in reload_plugin
  File "./pylinter.py", line 578, in <module>
    plugin_loaded()
  File "./pylinter.py", line 77, in plugin_loaded
    DEFAULT_PYLINT_COMMAND = PylSet.get_default_pylint_command()
  File "./pylinter.py", line 207, in get_default_pylint_command
    if not pylint_path:
UnboundLocalError: local variable 'pylint_path' referenced before assignment
```

Adding `pylint_path = None` on line 202 fixes this and the error dialogs appear... However, it doesn't seem to be picking up the values from config reliably. Perhaps related to #46.
My user config is

``` pre
{
    "python_path": "/Users/user/anaconda/lib/python2.7/",
    "pylint_path": "/Users/user/anaconda/lib/python2.7/site-packages/pylint/lint.py"
}
```

Before I get the "could not locate pylint message, I get",  "Fatal pylint error: ImportError: No module named pylint.checkers". This is importable from the command line.

``` pre
$ python
Python 2.7.6 |Anaconda 1.8.0 (x86_64)| (default, Jan 10 2014, 11:23:15) 
[GCC 4.0.1 (Apple Inc. build 5493)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pylint
>>> import pylint.checkers
>>> 
```

update: accidentally submitted comment.

**Comment by jlujan** [comment]
I can confirm that the pull request in #46 fixes my issue. In addition I had to specify python_bin in my settings file.

**Comment by biermeester** [comment]
I've finally merged #46. Thanks for the feedback!

**Comment by ankerfeng** [comment]
are you talking in sublimeText3 problems?
